### PR TITLE
Fix nil redis array element

### LINKED ISSUE #18765: [Release] actor_death is failing in weekly release test
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

```
(pid=gcs_server) [2021-09-20 03:50:23,732 C 515 515] redis_context.cc:157:  Check failed: REDIS_REPLY_STRING == entry->type Unexcepted type: 4
(pid=gcs_server) *** StackTrace Information ***
(pid=gcs_server)     ray::SpdLogMessage::Flush()
(pid=gcs_server)     ray::RayLog::~RayLog()
(pid=gcs_server)     ray::gcs::CallbackReply::ParseAsStringArray()
(pid=gcs_server)     ray::gcs::CallbackReply::CallbackReply()
(pid=gcs_server)     ray::gcs::GlobalRedisCallback()
(pid=gcs_server)     redisProcessCallbacks
(pid=gcs_server)     ray::gcs::RedisAsyncContext::RedisAsyncHandleRead()
(pid=gcs_server)     RedisAsioClient::handle_io()
(pid=gcs_server)     boost::asio::detail::executor_function<>::do_complete()
(pid=gcs_server)     boost::asio::io_context::executor_type::dispatch<>()
(pid=gcs_server)     boost::asio::detail::reactive_null_buffers_op<>::do_complete()
(pid=gcs_server)     boost::asio::detail::scheduler::do_run_one()
(pid=gcs_server)     boost::asio::detail::scheduler::run()
(pid=gcs_server)     boost::asio::io_context::run()
(pid=gcs_server)     main
(pid=gcs_server)     __libc_start_main
(pid=gcs_server) 
```

```
2021-09-20 03:50:29,229 WARNING worker.py:1227 -- The autoscaler failed with the following error:
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 317, in run
    self._run()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 207, in _run
    self.update_load_metrics()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 170, in update_load_metrics
    request, timeout=4)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "failed to connect to all addresses"
        debug_error_string = "{"created":"@1632135029.218620788","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3008,"referenced_errors":[{"created":"@1632135029.218615637","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":397,"grpc_status":14}]}"
>
```

### Reproduction script

https://buildkite.com/ray-project/periodic-ci/builds/1058#926de814-2e5b-4fca-8e03-51140a7ff58f

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
When we try to get all key-values from redis, we do `scan` first and then `mget`. Since they are not executed atomically, it's possible that some key returned from `scan` is deleted before `mget` so `mget` will return `nil`, this PR makes sure that we handle `nil` correctly instead of check failure.

## Related issue number
Closes #18765

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
Kicked off a new actor_deaths test based on this PR: https://buildkite.com/ray-project/periodic-ci/builds/1094#_.

**Review by wuisawesome — APPROVED** [review]
Sweet, let's make sure it passes the release test and then merge it.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
LINT failing: https://buildkite.com/ray-project/ray-builders-pr/builds/14364#d0772d3e-1ea2-4bdd-ac80-ae813c3bf42e

**Comment by jjyao** [comment]
The test passed: https://buildkite.com/ray-project/periodic-ci/builds/1101#076ac752-25d0-4c84-9403-868e71784d2b

**Comment by ericl** [comment]
Windows test looks flaky

File: src/ray/gcs/redis_context.cc [hunk 1]
```diff
@@ -183,11 +188,16 @@ const std::string &CallbackReply::ReadAsPubsubData() const {
...
+  array->reserve(string_array_reply_.size());
+  for (const auto &element : string_array_reply_) {
+    RAY_CHECK(element.has_value());
+    array->emplace_back(*element);
+  }
   return next_scan_cursor_reply_;
 }
 
-const std::vector<std::string> &CallbackReply::ReadAsStringArray() const {
+const std::vector<std::optional<std::string>> &CallbackReply::ReadAsStringArray() const {
```
Comment (mattip) [inline src/ray/gcs/redis_context.cc:200]: This change requires `std=c++17` and breaks my local build for MSVC on windows. Where does one specify to bazel that c++17 is required?

Comment (jjyao) [reply src/ray/gcs/redis_context.cc:200]: Hi @mattip, we have upgraded to c++17 recently: https://github.com/ray-project/ray/pull/18622/files

Comment (mattip) [reply src/ray/gcs/redis_context.cc:200]: Hmm. I wonder why building with `pip wheel .\python --use-feature=in-tree-build -vv` does not pick those changes up?

Comment (jjyao) [reply src/ray/gcs/redis_context.cc:200]: I'm actually not familiar with that part.  Can you open an issue if you see a problem? Thanks.

Comment (mattip) [reply src/ray/gcs/redis_context.cc:200]: #19070

**Pull request merged**